docs/GUIDELINES.md

```diff
@@ -0,0 +1,172 @@
+
+All outputs must have a `description` set. The `description` should be based on (or adapted from) the upstream terraform provider where applicable. Avoid simply repeating the variable name as the output `description`.
+
+## Use well-formatted snake case output names
+
+Avoid introducing any other syntaxes commonly found in other languages such as CamelCase or pascalCase. For consistency we want all variables to look uniform. It also makes code more consistent when using outputs together with terraform [`remote_state`](https://www.terraform.io/docs/providers/terraform/d/remote_state.html) to access those settings from across modules.
+
+## Use symmetrical names
+
+We prefer to keep terraform outputs symmetrical as much as possible with the upstream resource or module, with exception of prefixes. This reduces the amount of entropy in the code or possible ambiguity, while increasing consistency. Below is an example of what **not* to do. The expected output name is `user_secret_access_key`. This is because the other IAM user outputs in the upstream module are prefixed with `user_`, and then we should borrow the upstream’s output name of `secret_access_key` to become `user_secret_access_key` for consistency.
+
+# Naming Conventions
+
+## Use a programmatically consistent naming convention
+
+All resource names (E.g. things provisioned on AWS) must follow a consistent convention. The reason this is so important is that modules are frequently composed inside of other modules. Enforcing consistency increases the likelihood that modules can invoke other modules without colliding on resource names.
+
+To enforce consistency, we require that all modules use the [`terraform-null-label`](https://github.com/cloudposse/terraform-null-label) module. With this module, users have the ability to change the way resource names are generated such as by changing the order of parameters or the delimiter. While the module is opinionated on the parameters, it’s proved invaluable as a mechanism for generating consistent resource names.
+
+# Module Design
+
+## Small Opinionated Modules
+
+We believe that modules should do one thing very well. But in order to do that, it requires being opinionated on the design. Simply wrapping terraform resources for the purposes of modularising code is not that helpful. Implementing a specific use-case of those resource is more helpful.
+
+## Composable Modules
+
+Write all modules to be easily composable into other modules. This is how we’re able to achieve economies of scale and stop re-inventing the same patterns over and over again.
+
+## Use `variable` inputs
+
+Modules should accept as many parameters as possible. Avoid using inputs of `type = object` since they are harder to document. Of course, this is not a hard rule and sometimes objects just make the most sense. Just be weary of the ability for tools like [`terraform-docs`](https://github.com/segmentio/terraform-docs) to be able to generate meaningful documentation.
```

Comment:
```suggestion
Modules should accept as many parameters as possible. Avoid using inputs of `type = object` since they are harder to document. Of course, this is not a hard rule and sometimes objects just make the most sense. Just be aware of the ability for tools like [`terraform-docs`](https://github.com/segmentio/terraform-docs) to be able to generate meaningful documentation.
```